### ISSUE TITLE
TEST_CASE_CLASS requires explicit template instantiation

### ISSUE THREAD

**Issue by avoirtech** [doctest/doctest#718]
## Description
A test defined the TEST_CASE_CLASS within a template class will not be compiled or run unless the extra code is added outside the TEST_CASE_CLASS to explicitly instantiate the template

### Steps to reproduce
With the following two minimal examples in the same .h file, the first runs (failing as expected), the second does not run at all:

```c++
class NonTemplated {
    int foo;
public:
    TEST_CASE_CLASS("NonTemplatedRuns") {
        NonTemplated x;
        x.foo = 3;
        CHECK(x.foo == 10);
    }
};


template<typename T>
class Templated {
    T foo;
public:
    TEST_CASE_CLASS("TemplatedDoesNotRun") {
        Templated<int> x;
        x.foo = 3;
        CHECK(x.foo == 10);
    }
};
```

In order to get the templated example to run, it is necessary to do add something like this below:

```c++
template class Templated<int>;
```

Thanks to @Saalvage for pointing this out, but he also make the point [here](https://github.com/doctest/doctest/issues/702#issuecomment-1282559988) that it would run the same test for every instantiation present in your code, which may cause additional problems for some use cases.

### Extra information

* Instantiating the template class worked for me, though it does involve modifying production code, and instantiating templates like this may cause others issues.
* It would be a good idea to clearly mention this issue and workaround in the TEST_CASE_CLASS  documentation
* For a full fix, I wonder if this could be solved with an alternative to TEST_CASE_CLASS, something like the suggestion [here](https://github.com/doctest/doctest/issues/76#issuecomment-808603422). I would certainly prefer to not to have to use the TEST_CASE_CLASS within my class body in the .h, as when developing lots of tests for a well used type, this means that a trivial change to a test can end up requiring almost all my project needs to be recompiled, which puts a real barrier in the way of creating tests.
* doctest version on the issue-608 fix branch (v2.4.9 + )

**Comment by cdeln** [comment]
I don't think this is fixable to be honest. As I have argued in the other threads related to TEST_CASE_CLASS I think it shouldn't have been implemented in the first place, there are better ways to deal with this.

I suggest we close as wont fix because its impossible to fix.

**Comment by datadiode** [comment]
Working example:
```C++
template<typename T>
class Templated {
    T foo;
    class DocTest;
    friend DocTest;
};

template<>
class Templated<int>::DocTest {
    TEST_CASE_CLASS("Templated<int>") {
        Templated<int> x;
        x.foo = 3;
        CHECK(x.foo == 10);
    }
};

template<>
class Templated<unsigned>::DocTest {
    TEST_CASE_CLASS("Templated<unsigned>") {
        Templated<unsigned> x;
        x.foo = 3;
        CHECK(x.foo == 10);
    }
};
```

**Comment by mitchgrout** [comment]
> Working example:
> 
> [...]

Sorta but not really; the main problem here is that:

```cpp
template <typename...>
struct object {
  TEST_CASE_CLASS("...") { }
};
```

...does not register any tests, since `template`'s will not trigger codegen / global ctors unless they're instantiated. The original resolution was to `template struct object<...>` it (i.e. explicit instantiation). What you've proposed is basically that, except specialized for a particular type `T` (albeit in this example, the cases are identical).

The best fixes in my book would be:

1. Write the test out-of-class, using a `friend class` (like in your example) to access private members as-required
2. Write the test in-line (but out-of-class) with `TEST_CASE_TEMPLATE` to control the exact instantiations
3. Keep the test in-class, but `template struct object<...>` to explicitly register tests

For a header-only library, I'd pick the first choice, with tests being implemented in a dedicated test source file, since making `doctest` a mandatory dependency for consumers is a bit grotty (unless you use a bunch of `#ifdef` to avoid `#include <doctest.h>` thereby making it optional, but that's equally as gross).

**Comment by datadiode** [comment]
I didn't mean to offer a resolution which avoids explicit instantiation.  
If you try to `template class Templated<int>` then `template class Templated<char>` then only one of them will run its test.  
This is where my example makes a difference.  
Btw, the original repro code doesn't repro for me.  
Fixed repro code:
```C++
template<typename T>
class Templated {
    T foo;
    TEST_CASE_CLASS("TemplatedDoesNotRun") {
        Templated<T> x;
        x.foo = 3;
        CHECK(x.foo == 8);
    }
};
```
